### PR TITLE
Add fish queries

### LINKED ISSUE #1268: Add support for fish
**Language informations**

The only remaining language for me that doesn't have a treesitter parser is fish. I found a treesitter project for fish here:
https://github.com/krnik/tree-sitter-fish

However, it just gives me errors and I am not knowledgeable enough about treesitter to be able to understand why. I have created a seperate issue on the respective repository.

### PR DESCRIPTION
**Pull request by ram02z**
Closes #1268.
One small issue I have which I have not been able to figure out is how to highlight all command options like `--description`. As you can see from the highlight snippet below, only the first command option is highlighted (`--argument-names`) and following option on the same line (`--description`) is ignored. The following syntax highlight rule is from line 93:

```
(command
  argument: [
             (word) @parameter (#vim-match? @parameter "^-")
            ]
)
```

> Highlight

image

> Indent

Works well for every statement listed in `indents.scm` apart from `if_statement`. When adding `else` or `else if` clauses the indenation doesn't work and the `end` keyword indents on enter. See gif below.
if_indent_bug

> Folds

fish_foldr

### CODE REVIEW

**Comment by krnik** [comment]
> One small issue I have which I have not been able to figure out is how to highlight all command options like `--description`. As you can see from the highlight snippet below, only the first command option is highlighted (`--argument-names`) and following option on the same line (`--description`) is ignored. The following syntax highlight rule is from line 93:
> 
> ```
> (command
>   argument: [
>              (word) @parameter (#vim-match? @parameter "^-")
>             ]
> )
> ```

Fixed! I made a mistake when defining `command` rule.

**Comment by ram02z** [comment]
> [Fixed](https://github.com/krnik/tree-sitter-fish/commit/a83dad3344d640e2543b9df45ea8132ddcc581b9)! I made a mistake when defining `command` rule.

@krnik It still doesn't work on the function block, but works on other commands. I realised that line 93 was not for the function block since `function_definition` isn't a command😅.
I have also updated the `function_definition` highlight in https://github.com/ram02z/nvim-treesitter/commit/e90354b3ab2695196893bc119e4956823445e683 but I think the grammer needs an update also. 

image

File: queries/fish/highlights.scm [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+  option: [
+          (word)
+          (concatenation (word))
+          ] @parameter (#match? @parameter "^-")
+)
+
+;; Switch statements
+
+(switch_statement
+    value: (word) @parameter)
```
Comment (vigoux) [inline queries/fish/highlights.scm:128]: Not sure the `parameter` is right here, as it is not exactly the case.
This should be something like `constant`.

File: queries/fish/highlights.scm [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+
+(case_clause ["case"] @keyword.function)
```
Comment (vigoux) [inline queries/fish/highlights.scm:130]: Here too, this is not a function related keyword.

**Review by vigoux — COMMENTED** [review]
Awesome, thanks for handling that!!
If you want you can also add you as the fish maintainer in the repo!

I left a few comments btw.

**Comment by krnik** [comment]
@ram02z Added fixes for other incorrect fields (function and for loop)

**Comment by krnik** [comment]
Let's also add the query `(command name: (word) @function)` to highlight non-builtin command names.
Like in bash https://github.com/nvim-treesitter/nvim-treesitter/blob/master/queries/bash/highlights.scm#L90

**Comment by vigoux** [comment]
I think what you should do is rebase on #1269 just to get the parser set up, and that'll make the CI run on your addition.

File: queries/fish/highlights.scm (hunk 1 above)
Comment (ram02z) [reply queries/fish/highlights.scm]: I removed this. Referenced languages like golang and bash and none of them highlight the case value.

File: queries/fish/highlights.scm (hunk 2 above)
Comment (ram02z) [reply queries/fish/highlights.scm]: Moved `case_clause` to `@conditional`.

**Comment by ram02z** [comment]
> 
> Let's also add the query `(command name: (word) @function)` to highlight non-builtin command names.
> Like in bash https://github.com/nvim-treesitter/nvim-treesitter/blob/master/queries/bash/highlights.scm#L90

Added this along with the changes requested by @vigoux .

**Comment by ram02z** [comment]
> 
> 
> I think what you should do is rebase on #1269 just to get the parser set up, and that'll make the CI run on your addition.

I rebased on #1269. Not sure why it added 2873d1efdaedf562af202d13594f1022def1fa49 along with the commits but I hope that is still fine.

File: lua/nvim-treesitter/parsers.lua [hunk 3]
```diff
@@ -140,6 +140,13 @@ list.bash = {
   maintainers = {"@TravonteD"},
 }
 
+list.fish = {
+  install_info = {
+    url = "https://github.com/krnik/tree-sitter-fish",
+    files = { "src/parser.c", "src/scanner.c" },
+  },
+}
+
```
Comment (vigoux) [inline lua/nvim-treesitter/parsers.lua:149]: If you want to maintain the queries for fish, you should add your name here in the `maintainers` key (see other entries in the parser list for examples).

**Review by vigoux — APPROVED** [review]

File: lua/nvim-treesitter/parsers.lua (hunk 3 above)
Comment (ram02z) [reply lua/nvim-treesitter/parsers.lua:149]: Will do! @krnik Should I add your name as well?

Comment (krnik) [reply lua/nvim-treesitter/parsers.lua:149]: I'll be glad to help if something breaks so yes, please add me too.

**Comment by ram02z** [comment]
Added the maintainer names for the fish parser list.

**Comment by kidonng** [comment]
The README seems to need updating as well.

**Comment by vigoux** [comment]
Hmmm yep it should be... Maybe in the next README update, I'll check that !

**Pull request merged**